cpp/src/io/parquet/writer_impl.cu

```diff
@@ -1407,17 +1453,60 @@ void writer::impl::write(table_view const& table, std::vector<partition_info> co
     });
 
   // Init page fragments
-  // 5000 is good enough for up to ~200-character strings. Longer strings will start producing
-  // fragments larger than the desired page size -> TODO: keep track of the max fragment size, and
-  // iteratively reduce this value if the largest fragment exceeds the max page size limit (we
-  // ideally want the page size to be below 1MB so as to have enough pages to get good
-  // compression/decompression performance).
-  // If using the default fragment size, scale it up or down depending on the requested page size.
-  if (max_page_fragment_size_ == cudf::io::default_max_page_fragment_size) {
-    max_page_fragment_size_ = (cudf::io::default_max_page_fragment_size * max_page_size_bytes) /
-                              cudf::io::default_max_page_size_bytes;
+  // 5000 is good enough for up to ~200-character strings. Longer strings and deeply nested columns
+  // will start producing fragments larger than the desired page size, so calculate fragment sizes
+  // for each leaf column.  Skip if the fragment size is not the default.
+  bool is_default_fragment_size = max_page_fragment_size_ == -1;
+  if (is_default_fragment_size) { max_page_fragment_size_ = default_max_page_fragment_size; }
+
+  std::vector<size_type> column_frag_size;
+
+  if (table.num_rows() > 0 && is_default_fragment_size) {
+    std::vector<size_t> column_sizes;
+    std::transform(single_streams_table.begin(),
+                   single_streams_table.end(),
+                   std::back_inserter(column_sizes),
+                   [this](auto const& column) { return column_size(column, stream); });
+
+    // adjust global fragment size if a single fragment will overrun a rowgroup
+    auto const table_size  = std::reduce(column_sizes.begin(), column_sizes.end());
+    auto const avg_row_len = util::div_rounding_up_safe<size_t>(table_size, table.num_rows());
+    if (avg_row_len > 0) {
+      auto const rg_frag_size = util::div_rounding_up_safe(max_row_group_size, avg_row_len);
+      max_page_fragment_size_ = std::min<size_type>(rg_frag_size, max_page_fragment_size_);
+    }
+
+    // dividing page size by average row length will tend to overshoot the desired
+    // page size when there's high variability in the row lengths. instead, shoot
+    // for multiple fragments per page to smooth things out. using 2 was too
+    // unbalanced in final page sizes, so using 4 which seems to be a good
+    // compromise at smoothing things out without getting fragment sizes too small.
+    auto frag_size_fn = [&](auto const& col, size_type col_size) {
+      const int target_frags_per_page = is_fixed_width(col.type()) ? 1 : 4;
```

Comment:
how do structs fit in here? a struct containing only fixed-width fields is effectively a fixed-width type.

Reply:
True...I'll add a function that tests for that.  Fixed length byte arrays too when they're supported.